Add small pr guidelines (#2977)
GoogleContainerTools/skaffold: docs/community/small-prs.md

# Pull Request Guidelines

> Writing good pull requests increase development velocity and minimize frustration.

Reviewer: increases and minimizes?

> If pull requests are small they will result in being opened and merged quickly.

Reviewer 2: ```suggestion
Small pull request will be merged quicker.
```

Having big pull requests that bounce back and forth between developers
and reviewers can slow progress down significantly, causing developers to waste a 
lot of time dealing with merge conflicts.

Big pull requests also increase the risk of breaking things; since changes are so big, 

>  - it’s hard to scrutinize all the changes and 

Reviewer 2: ```suggestion
 - it’s hard to truly understand all the changes and 
```

 - test them for regressions.

In order to promote good PR quality, the skaffold team will push back
against big PRs. See [0](https://github.com/GoogleContainerTools/skaffold/pull/2750#pullrequestreview-283621241), [1](https://github.com/GoogleContainerTools/skaffold/pull/2917#pullrequestreview-291415741)

## Breaking down Pull Requests

> ### Feature development.

Reviewer 2: ```suggestion
### Feature development
```

Adding a new command or skaffold config results in large number of changes due to
all the boilerplate code required for adding a new command, generating docs etc.

It makes sense to break down these changes into 2 categories:

> 1. Incremental code changes which are not accessed by control flow.

Reviewer 2: ```suggestion
1. "Invisible changes": Incremental, small code changes 
     1. refactoring: either keep the functionality the same but refactoring the design or
     1. partial implementations: unit tested, new behavior, that is not yet accessible by the main control flow.
```

> 2. User affecting changes like

Reviewer 2: ```suggestion
2. "Visible changes": User affecting changes, for example
```

   1. Adding a new command which exercises the above code changes
   1. Adding config change and generating docs.

#### Config Change Example   
Below is an example of how you can break a config change into small individual PRs.

A contributor wanted to support [custom build args to `kustomize` deployer](https://github.com/GoogleContainerTools/skaffold/issues/2488).

> This required, skaffold to support a `buildArgs` field to `deploy.kustomize` config. Plumb the config to `kustomize` command.

Reviewer 2: ```suggestion
This required skaffold to add the `deploy.kustomize.buildArgs` field and then plumb the arguments to the `kustomize` command.
```

1. Add this feature to `kubectl`, `helm` and `kustomize` deployers supported by skaffold.
2. support a new flag for `skaffold deploy` and `skaffold run` command
3. may be add a new command to only render k8 manifests.

Item 2 and 3 in the above list of changes, are user facing features and could be abstracted out.
Item 1 can further be implemented incrementally by supporting this new feature for each of the deployer.

To implement this feature, we 
1. First added an [method to interface `deployer` with empty stub](https://github.com/GoogleContainerTools/skaffold/pull/2834).
   
   Note: The code added in this PR does not get exercised other than in tests.
2. Then add render implementation for each [kubectl deployer](https://github.com/GoogleContainerTools/skaffold/pull/2943), kustomize and helm.
3. Add a [render command](https://github.com/GoogleContainerTools/skaffold/pull/2942)

Remember, its ok to land code which is not exercised. However please mention the follow up work
in **Next PRs** section when creating a Pull Request [e.g. here.](https://github.com/GoogleContainerTools/skaffold/pull/2811)

> Finally, please use our best judgement when submitting pull requests.

Reviewer 2: ```suggestion
Finally, please use your best judgement when submitting pull requests, these rules might not always work for you - we would love to hear that! 
```
